pkg/pg/txdb.go

```diff
@@ -0,0 +1,531 @@
+// negative impact on normal test operation.
+//
+// If you MUST test BEGIN/ROLLBACK behaviour, you will have to configure your
+// store to use the raw DialectPostgres dialect and setup a one-use database.
+// See heavyweight.FullTestDB() as a convenience function to help you do this,
+// but please use sparingly because as it's name implies, it is expensive.
+func RegisterTxDb(dbURL string) error {
+	drivers := sql.Drivers()
+	for _, driver := range drivers {
+		if driver == string(TransactionWrappedPostgres) {
+			// TxDB driver already registered
+			return nil
+		}
+	}
+	if testing.Short() {
+		// -short tests don't need a DB
+		return nil
+	}
+	parsed, err := url.Parse(dbURL)
+	if err != nil {
+		return fmt.Errorf("failed to parse %s as a database URL: %w", dbURL, err)
+	}
+	if parsed.Path == "" {
+		return fmt.Errorf("database url `%s` must point to your test database. Note that the test database MUST end in `_test` to differentiate from a possible production DB. HINT: Try postgresql://postgres@localhost:5432/chainlink_test?sslmode=disable", parsed.String())
+	}
+	if !strings.HasSuffix(parsed.Path, "_test") {
+		return fmt.Errorf("cannot run tests against database named `%s`. Note that the test database MUST end in `_test` to differentiate from a possible production DB. HINT: Try postgresql://postgres@localhost:5432/chainlink_test?sslmode=disable", parsed.Path[1:])
+	}
+
+	name := string(TransactionWrappedPostgres)
+	sql.Register(name, &txDriver{
+		abort: make(chan struct{}),
+		dbURL: dbURL,
+		conns: make(map[string]*conn),
+	})
+	sqlx.BindDriver(name, sqlx.DOLLAR)
+	return nil
+}
+
+var _ driver.Conn = &conn{}
+
+var _ driver.Validator = &conn{}
+var _ driver.SessionResetter = &conn{}
+
+// txDriver is a sql driver which runs on a single transaction.
+// When `Close` is called, transaction is rolled back.
+type txDriver struct {
+	sync.Mutex
+	abort chan struct{}
+	db    *sql.DB
+	conns map[string]*conn
+
+	dbURL string
+}
+
+func (d *txDriver) Open(dsn string) (driver.Conn, error) {
+	d.Lock()
+	defer d.Unlock()
+	// Open real db connection if its the first call
+	if d.db == nil {
+		db, err := sql.Open(string(Postgres), d.dbURL)
+		if err != nil {
+			return nil, err
+		}
+		d.db = db
+	}
+	c, exists := d.conns[dsn]
+	if !exists || !c.tryOpen() {
+		tx, err := d.db.Begin()
+		if err != nil {
+			return nil, err
+		}
+		c = &conn{abort: d.abort, tx: tx, opened: 1, dsn: dsn}
+		c.removeSelf = func() error {
+			return d.deleteConn(c)
+		}
+		d.conns[dsn] = c
+	}
+	return c, nil
+}
+
+// deleteConn is called by a connection when it is closed via the `close` method.
+// It also auto-closes the DB when the last checked out connection is closed.
+func (d *txDriver) deleteConn(c *conn) error {
+	// must lock here to avoid racing with Open
+	d.Lock()
+	defer d.Unlock()
+
+	if d.conns[c.dsn] != c {
+		return nil // already been replaced
+	}
+	delete(d.conns, c.dsn)
+	if len(d.conns) == 0 && d.db != nil {
+		close(d.abort)
```

Comment:
Good point, but the underlying problem was a bit more than that. Sharing only a single abort channel across all connections doesn't work since different tests close their connections at different times. I changed it to have only 1 abort channel per connection now--sort of.  The only exception is if two connections have the same dsn string passed to them, but it looks like we always generate them randomly when calling `Open()` so hopefully there can't be a conflict.